### ISSUE TITLE
Larger buttons in the UI

### ISSUE THREAD

**Issue by bunnybot** [widelands/widelands#1294]
A while ago I started thinking about the buttons in the UI, which are rather small heightwise. Accessibility wise this makes it harder to click them than it would if they were taller. They are already quite wide, so they are easy to select in the width, but slightly harder to select in height. Also, looking at the various menus, there is a lot of whitespace which is not used for anything. Rather than having lots of wasted screen area, I suggest rather filling it with larger buttons. This change would also increase accessibility as it gets easier to click the intended button when it has a larger click area.

I have experimented a bit with adjusting the height on the main menu, which makes the buttons a bit taller. However, from what I can see the width/height of the various buttons are set on each menu? Is there any reason this isn't defined in one place, to ensure all buttons behave the same. I was happy to see they automatically scale to the resolution set, though.

Since other changes to the user interface which might happen this iteration http://wl.widelands.org/forum/topic/724/ , I thought I'd post this early in the build17 cycle to get some feedback and discussion going. I think lots of the work posted in that forum thread looks great, so I suspect build17 might see some large changes to the look and feel of Widelands.

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2011-04-17T11:07:04Z
 * owner: hjd
 * assignee: shevonar
 * the launchpad url was https://bugs.launchpad.net/bugs/763517

**Comment by bunnybot** [comment]
(by hjd)
Just a demonstration. I played a bit around with various values for button height. 
This is *not* intended as a suggestion, more to better display my suggestion.

**Comment by bunnybot** [comment]
(by sirver)
A resonable suggestions imho. I think close colaboration with the graphics team is most important.

**Comment by bunnybot** [comment]
(by chuckw20)
Personally, I like this suggestion for a number of reasons and think it should be implemented.

**Comment by bunnybot** [comment]
(by hjd)
Ok, I have made some attempts to establish a common default height and width for buttons, but have so far been unsuccessful. The current situation is that each menu, for instance src/ui_fsmenu/main.cc:
        m_butw (get_w() * 7 / 20),
        m_buth (get_h() * 18 / 300),
These variables are then sent as arguments when the buttons are created. This means that the default height/width is duplicated in each file needing it, and all occurrences need to be updated. 

My suggestion is to organize so that it is possible to call the button constructor without specifying height/width and it will be assigned a default value. This would allow us to move it to one place where it can be updated. Also for any buttons we don't want the default size, calling the constructor with height/width parameters will override it. (In essence two constructors, one with height/width and one without, where the latter would redirect to the first with the default values specified) I have tried adding default values to the constructor, but the compiler does not like it when the rest of the parameters have no default values. 

However, I suspect the reason it is done the way it is currently is due to the get_w()/get_h() methods which will not be available in the button class. Which means my approach may not even work if done correctly. I would still really like to see this defined in a single place. Does anyone have any ideas?

Oh, and I'm not sure if I have mentioned this already, but while I know programming, I don't really know c++ :/

**Comment by bunnybot** [comment]
(by sirver)
those calculations are done to make the button's size resolution independent in the menus, defaulting to one button size would be bad if you increase the resolution: the buttons would need scaling.

A better approach might be remodel the UI to work with normalized coordinates going from 0 to 1 or so. With this, the resolution would not matter for the layout and your approach with the default sizes will work out.

As it stands now, I cannot see how the button default size can be moved inside the button class besides adding a static member variable. That is clutchy though.

**Comment by bunnybot** [comment]
(by hjd)
>those calculations are done to make the button's size resolution independent in the menus
yes, that's what I thought. 

I am not sure about your suggestion about normalized coordinates. I think that may make it harder to finetune positions. 

I thought a bit about making a variable (call it a semiconstant) which holds the current button-height/width scaled to the current resolution. This would be a "semiconstant" since it would only be updated when the resolution changes (singleton?). It seems a bit hacky though...

**Comment by bunnybot** [comment]
(by sirver)
thats along the lines what I suggested with a static member in the Buttons class. Not nice.

The normalized variables idea come from recent open gl based games that define their menus in such a way. Upping the resolution doesn't change the size of the elements on the screen, only their resolution. I find this really neat.

**Comment by bunnybot** [comment]
(by hjd)
>The normalized variables idea come from recent open gl based games that define their menus in such a way

Any examples I can take a look at? I'd like to see how it works before commenting on  it further.

**Comment by bunnybot** [comment]
(by sirver)
One example is Starcraft 2. I guess all newer 3D games will use this.

**Comment by bunnybot** [comment]
(by hjd)
This has been on my todo list for a while, but I guess we will just have to wait a while. 

I would like to be able to see how the tweaked buttons work over time, and allow time for adjustments both for improvement and consistency. Therefore I consider this a bit late in the release cycle, so I'm retargetting this to build18. I plan set up an experimental branch with this as soon as build17 is out the door to get this done, though.

**Comment by bunnybot** [comment]
(by joe-nahmias)
Hmm, this seems to dovetail with Bug #1030691 that I just opened.  It would be nice if that one made it into M-18 as well.

**Comment by bunnybot** [comment]
(by chuckw20)
FYI - in case it might be convenient as a test bed, I have updated the GUI-Development public branch (https://code.launchpad.net/~widelands-dev/widelands/GUI-Development) on launchpad with the latest revisions in the trunk.

**Comment by bunnybot** [comment]
(by sirver)
not sure what the status of this is. But it seems not be happening for b18.

**Comment by bunnybot** [comment]
(by sirver)
Setting to incomplete for bug sweeping.

**Comment by bunnybot** [comment]
(by gunchleoc)
IMO The easiest way to have a common button height is to define it in the constructor of the FullscreenMenuBase and have it as a protected const variable. That's the way I'm going at the moment anyway.

We also have buttons in the wui, which are of different sized to the fs_menu buttons, so I don't think a general solution in the UI::Button class would be viable.

**Comment by schnedann** [comment]
Just a comment of somebody who just installed widelands and has a 4k display. Scaling is a must have now. Everything is way too small. Icons, especially in the headqarter menue, and all fonts are way to small, I even need to go nearer to the monitor to read it. And you already guessed it: there is no difference in fullscreen and windowed - the scale is the same in both versions. It also seems setting a resolution in fullscreen does not work (OK thats what "windowssize" implies). It always uses what the monitor offers. At least render fullscreen with a lower resolution should be a viable solution without hopefully change to much code in first place...
For now my conclusion on widelands on my display is "unplayable"

**Comment by eveDude** [comment]
Playing on 1400p over here, while most of the stuff is okay-ish, the menus (like action menu) are really really tiny.
Would be neat if there was some (independant) UI scaling.

Hijacking this issue, since it came up when I googled..

**Comment by bunnybot** [comment]
[<img align='left' src='https://codeberg.org/avatars/4ff662299c00e9020397c275f41f0102' alt='Zeronior' height='48' />](https://codeberg.org/Zeronior)*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/1294#issuecomment-15007740)*
*On Fri May 15 16:18:24 CEST 2026, [**Zeronior**](https://codeberg.org/Zeronior) wrote:*

---

First, I have to start Widelands with the terminal. Otherwise, it's just a tiny part of the whole screen, in the left bottom corner.
Then, I have to use ALT+Scrollwheel to let Cinnamon zoom into the game.
If I choose to set the UI scaling to something about 2.5x, texts almost get readable while buttons are much too tiny to decipher (without Cinnamon zoom).
There are some older extensions, but none work as they should.
They just do the same as the GUI scaling: breaking the GUI.

Whatever I do, the game gets unplayable. The large fonts in the screenshots are my "normal" reading size: About 16 px.

**Comment by bunnybot** [comment]
[<img align='left' src='https://codeberg.org/avatars/593dc43e2444c25db9481960e5b693ab' alt='tothxa' height='48' />](https://codeberg.org/tothxa)*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/1294#issuecomment-15024984)*
*On Fri May 15 16:46:20 CEST 2026, [**Tóth András (tothxa)**](https://codeberg.org/tothxa) wrote:*

---

<@>Zeronior Please be aware that UI scaling is currently in development. Only the first steps had been done yet, there's still much to do before it gets really usable.

**Comment by bunnybot** [comment]
[<img align='left' src='https://codeberg.org/avatars/4ff662299c00e9020397c275f41f0102' alt='Zeronior' height='48' />](https://codeberg.org/Zeronior)*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/1294#issuecomment-15029163)*
*On Fri May 15 18:47:27 CEST 2026, [**Zeronior**](https://codeberg.org/Zeronior) wrote:*

---

It's happened with version 1.1, 1.3 and 1.4
